examples/cfq/utils.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utils module"""
+import os
+import pickle
+from typing import Iterable, List, Sequence, Dict, Text, Any
+import collections
+
+import numpy as np
+import tensorflow_text as text
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+from absl import logging
+
+import constants
+
+
+def _get_tokens(input_features: Dict, 
+                tokenizer: text.Tokenizer,
+                datasets: Iterable[tf.data.Dataset],
+                preprocessing_fun: Any) -> Iterable[List[bytes]]:
+    """Returns a list of tokens for all input fields in the given datasets."""
+
+    def _tokenize_input_features(
+            example: Dict[Text, tf.Tensor]) -> Dict[Text, tf.Tensor]:
+        """Tokenizes all input features in an example."""
+        example = preprocessing_fun(example)
+        for feature in example:
+            if feature in input_features:
+                example[feature] = tokenizer.tokenize(example[feature])
+        return example
+
+    for dataset in datasets:
+        # Apply the tokenizer to all input features.
+        tokenized_dataset = dataset.map(_tokenize_input_features,
+                                        num_parallel_calls=constants.AUTOTUNE)
+        # Yield all tokenized input features (i.e., tokenized input sentences).
+        for example in tfds.as_numpy(tokenized_dataset):
+            for feature in input_features:
+                yield example[feature]
+
+
+def build_vocabulary(file_name: Text,
+                     input_features: Dict,
+                     tokenizer: text.Tokenizer,
+                     datasets: Iterable[tf.data.Dataset],
+                     special_tokens: Sequence[bytes] = (constants.PAD,
+                                                        constants.UNK,
+                                                        constants.BOS,
+                                                        constants.EOS),
+                     preprocessing_fun: Any = lambda x: x) -> Dict[bytes, int]:
```

Comment:
You can use Callable from typing to specify the type of a function: https://docs.python.org/3/library/typing.html#callable